fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/AggregatedMaterializedViewRewriter.java

```diff
@@ -106,190 +106,140 @@ public boolean isValidPlan(OptExpression expression) {
+        ColumnRewriter columnRewriter = new ColumnRewriter(rewriteContext);
+        List<ScalarOperator> mvGroupingKeys =
+                rewriteGroupByKeys(rewriteContext, columnRewriter, mvAggOp.getGroupingKeys(), true);
+        List<ScalarOperator> queryGroupingKeys =
+                rewriteGroupByKeys(rewriteContext, columnRewriter, queryAggOp.getGroupingKeys(), false);
+        boolean isRollup = isRollupAggregate(mvGroupingKeys, queryGroupingKeys);
+
+        // Cannot ROLLUP distinct
+        if (isRollup && mvAggOp.getAggregations().values().stream().anyMatch(callOp -> callOp.isDistinct())) {
             return null;
         }
 
-        // check aggregates of query
         // normalize mv's aggs by using query's table ref and query ec
         Map<ColumnRefOperator, ScalarOperator> mvProjection =
                 MvUtils.getColumnRefMap(rewriteContext.getMvExpression(), rewriteContext.getMvRefFactory());
         // normalize view projection by query relation and ec
         Multimap<ScalarOperator, ColumnRefOperator> normalizedViewMap =
                 normalizeAndReverseProjection(mvProjection, rewriteContext, false);
 
-
         if (isRollup) {
-            return rewriteForRollup(queryAgg, swappedQueryAggs, swappedMvGroupingKeys,
-                    queryAggGroupingKeys, normalizedViewMap, rewriteContext, targetExpr);
+            Map<ColumnRefOperator, ScalarOperator> queryColumnRefToAggMap = rewriteAggregations(
+                    rewriteContext, columnRewriter, queryAggOp.getAggregations(), false);
+            return rewriteForRollup(queryAggOp, queryColumnRefToAggMap, queryGroupingKeys,
+                    normalizedViewMap, rewriteContext, mvOptExpr);
         } else {
-            return rewriteProjection(rewriteContext, normalizedViewMap, targetExpr);
+            return rewriteProjection(rewriteContext, normalizedViewMap, mvOptExpr);
         }
     }
 
-    // judge whether the output column is a projection on grouping key
-    private boolean isBasedOnKey(List<ScalarOperator> keys, ScalarOperator targetScalar) {
-        ScalarOperatorVisitor<Boolean, Void> keyJudger = new ScalarOperatorVisitor<Boolean, Void>() {
-            @Override
-            public Boolean visit(ScalarOperator scalarOperator, Void context) {
-                for (ScalarOperator child : scalarOperator.getChildren()) {
-                    Boolean ret = child.accept(this, context);
-                    if (ret) {
-                        return true;
-                    }
+    // NOTE: this method is not exactly right to check whether it's a rollup aggregate:
+    // - all matched group by keys bit is less than mvGroupByKeys
+    // - if query contains one non-mv-existed agg, set it `rollup` and use `replaceExprWithTarget` to
+    //    - check whether to rewrite later.
+    private boolean isRollupAggregate(List<ScalarOperator> mvGroupingKeys, List<ScalarOperator> queryGroupingKeys) {
```

Comment:
how to support these cases?
1. mv group by a, b; query: group by date_trunc(a, "month"), b
2. mv group by a, b, query : group by a * b?